File: Book/C/Fig_12.21_ompTwoTarg.c
#include <omp.h>
#include <stdlib.h>
#include <stdio.h>
#define N 1024
int main()
{
   float *a, *b, *c, *d;
   int i;
   
   a = (float*) malloc(N * sizeof(float));
   b = (float*) malloc(N * sizeof(float));   
   c = (float*) malloc(N * sizeof(float));
   d = (float*) malloc(N * sizeof(float));
   
// initialize a, b, c, and d (code not shown)

#pragma omp target map(to:a[0:N],b[0:N]) map(tofrom:c[0:N])
#pragma omp teams distribute parallel for simd
   for (i = 0; i < N;i++)
      c[i] += a[i] * b[i];
     
#pragma omp target map(to:a[0:N],c[0:N]) map(tofrom:d[0:N])
#pragma omp teams distribute parallel for simd
   for (i = 0; i < N; i++)
      d[i] += a[i] + c[i];
}

File: Book/Fortran/Fig_12.21_ompTwoTarg.f90
program main
   use omp_lib

   integer, parameter :: N = 1024
   real*8, allocatable, dimension (:) :: a, b, c, d
   integer :: i

   allocate (a(N))
   allocate (b(N))
   allocate (c(N))
   allocate (d(N))

! initialize a, b, and c (code not shown)

!$omp target map(to:a(1:N),b(1:N)) map(tofrom:c(1:N))
!$omp teams distribute parallel do simd
   do i = 1, N
      c(i) = c(i) + a(i) * b(i)
   enddo
!$omp end teams distribute parallel do simd
!$omp end target

!$omp target map(to:a(1:N),b(1:N)) map(tofrom:d(1:N))
!$omp teams distribute parallel do simd
   do i = 1, N
      d(i) = d(i) + a(i) * c(i)
   enddo
!$omp end teams distribute parallel do simd
!$omp end target

end program main
